src/character.cpp

```diff
@@ -4046,7 +4046,9 @@ void Character::mod_stored_calories( int ncal, const bool ignore_weariness )
     if( !ignore_weariness ) {
         activity_history.calorie_adjust( ncal );
     }
-    set_stored_calories( stored_calories + ncal );
+    if( ncal < 0 || ncal + stored_calories > 0 ) {
+        set_stored_calories( stored_calories + ncal );
+    }
```

Comment:
```suggestion
    static_assert( std::is_same<decltype( ncal ), decltype( stored_calories )>::value, "Expected the same type for ncal and stored_calories" );
    const decltype( ncal ) cal_max = std::numeric_limits<decltype( ncal )>::highest();
    const decltype( ncal ) cal_min = std::numeric_limits<decltype( ncal )>::lowest();
    if( ( ncal <= 0 && stored_calories >= 0 ) || ( ncal >= 0 && stored_calories <= 0 ) ) {
        // non-negative + non-positive must be no less than the lowest negative and no more than the highest positive and therefore never overflows
        set_stored_calories( stored_calories + ncal );
    } else if( ncal > 0 ) {
        // therefore stored_calories > 0
        if( cal_max - stored_calories >= ncal ) {
            set_stored_calories( stored_calories + ncal );
        } else {
            set_stored_calories( cal_max );
        }
    } else {
        // ncal < 0 and stored_calories < 0
        if( cal_min - stored_calories <= ncal ) {
            set_stored_calories( stored_calories + ncal );
        } else {
            set_stored_calories( cal_min );
        }
    }
```
This might look intimidating but I think it is the safest approach. I haven't compiled and tested it though.

Reply:
Alright, I'll hold on for a senior dev on this.

Reply:
I think adding negatives isn't really relevant here since `stored_calories <= 0` means starving to death, so doing all of this might be a bit overkill. Although it might be useful to make it into a general "safe addition" helper function. I'm sure we have other places in the code with custom overflow guards.

Reply:
This might be appropriate for a generic helper for "does_sum_overflow()", but is massively invertebrates for inlining.